File: service/djamp_service/core/certificate.py
"""
Certificate management module for DJANGOForge
Handles Root CA and SSL certificate generation
"""

from pathlib import Path
from typing import Dict, Optional
import subprocess
from datetime import datetime


class CertificateManager:
    """Manages SSL certificates for local development"""

    def __init__(self):
        self.djamp_home = Path.home() / ".djamp"
        self.ca_dir = self.djamp_home / "ca"
        self.cert_dir = self.djamp_home / "certs"
        self.ca_cert_path = self.ca_dir / "djangoforge-root-ca.crt"
        self.ca_key_path = self.ca_dir / "djangoforge-root-ca.key"

    def _ensure_directories(self):
        """Ensure all required directories exist"""
        self.ca_dir.mkdir(parents=True, exist_ok=True)
        self.cert_dir.mkdir(parents=True, exist_ok=True)

    async def generate_root_ca(self) -> Dict[str, any]:
        """Generate Root CA certificate"""
        try:
            self._ensure_directories()

            # Check if CA already exists
            if self.ca_cert_path.exists() and self.ca_key_path.exists():
                return {"success": True, "message": "Root CA already exists"}

            # Generate Root CA
            process = await asyncio.create_subprocess_exec(
                "openssl",
                "req",
                "-x509",
                "-newkey",
                "rsa:4096",
                "-keyout",
                str(self.ca_key_path),
                "-out",
                str(self.ca_cert_path),
                "-days",
                "3650",
                "-nodes",
                "-subj",
                "/C=US/ST=State/L=City/O=DJANGOForge/OU=Development/CN=DJANGOForge Root CA",
                "-addext",
                "basicConstraints=critical,CA:TRUE,pathlen:0",
                "-addext",
                "keyUsage=critical,keyCertSign,cRLSign",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            stdout, stderr = await process.communicate()

            if process.returncode != 0:
                return {"success": False, "error": stderr.decode()}

            return {"success": True, "certificate_path": str(self.ca_cert_path)}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def generate_certificate(self, domain: str) -> Dict[str, any]:
        """Generate SSL certificate for a domain"""
        try:
            self._ensure_directories()

            # Check if Root CA exists
            if not self.ca_cert_path.exists() or not self.ca_key_path.exists():
                return {"success": False, "error": "Root CA not found. Generate Root CA first."}

            cert_key_path = self.cert_dir / f"{domain}.key"
            cert_crt_path = self.cert_dir / f"{domain}.crt"
            config_path = self.cert_dir / f"{domain}.conf"

            # Create config file
            config_content = f"""
[req]
distinguished_name = req_distinguished_name
req_extensions = v3_req
prompt = no

[req_distinguished_name]
CN = {domain}

[v3_req]
keyUsage = keyEncipherment, dataEncipherment
extendedKeyUsage = serverAuth
subjectAltName = @alt_names

[alt_names]
DNS.1 = {domain}
DNS.2 = www.{domain}
"""
            config_path.write_text(config_content)

            # Generate private key
            process = await asyncio.create_subprocess_exec(
                "openssl",
                "genrsa",
                "-out",
                str(cert_key_path),
                "2048",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await process.communicate()

            if process.returncode != 0:
                return {"success": False, "error": "Failed to generate private key"}

            # Generate CSR
            csr_path = self.cert_dir / f"{domain}.csr"
            process = await asyncio.create_subprocess_exec(
                "openssl",
                "req",
                "-new",
                "-key",
                str(cert_key_path),
                "-out",
                str(csr_path),
                "-config",
                str(config_path),
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await process.communicate()

            if process.returncode != 0:
                return {"success": False, "error": "Failed to generate CSR"}

            # Sign certificate
            process = await asyncio.create_subprocess_exec(
                "openssl",
                "x509",
                "-req",
                "-in",
                str(csr_path),
                "-CA",
                str(self.ca_cert_path),
                "-CAkey",
                str(self.ca_key_path),
                "-CAcreateserial",
                "-out",
                str(cert_crt_path),
                "-days",
                "365",
                "-sha256",
                "-extensions",
                "v3_req",
                "-extfile",
                str(config_path),
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await process.communicate()

            # Clean up
            csr_path.unlink(missing_ok=True)
            config_path.unlink(missing_ok=True)

            if process.returncode != 0:
                return {"success": False, "error": stderr.decode()}

            # Get expiration date
            expiration = await self._get_certificate_expiration(cert_crt_path)

            return {
                "success": True,
                "domain": domain,
                "certificate_path": str(cert_crt_path),
                "key_path": str(cert_key_path),
                "expires_at": expiration,
            }
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def _get_certificate_expiration(self, cert_path: Path) -> str:
        """Get certificate expiration date"""
        process = await asyncio.create_subprocess_exec(
            "openssl",
            "x509",
            "-enddate",
            "-noout",
            "-in",
            str(cert_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await process.communicate()

        if process.returncode == 0:
            return stdout.decode().replace("notAfter=", "").strip()
        return ""

    async def check_certificate_status(self, domain: str) -> Dict[str, any]:
        """Check if certificate exists and is valid"""
        cert_path = self.cert_dir / f"{domain}.crt"
        key_path = self.cert_dir / f"{domain}.key"

        if not cert_path.exists() or not key_path.exists():
            return {
                "domain": domain,
                "certificate_path": str(cert_path),
                "key_path": str(key_path),
                "is_valid": False,
            }

        # Check validity
        process = await asyncio.create_subprocess_exec(
            "openssl",
            "x509",
            "-checkend",
            "0",
            "-in",
            str(cert_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()

        is_valid = process.returncode == 0

        expiration = await self._get_certificate_expiration(cert_path)

        return {
            "domain": domain,
            "certificate_path": str(cert_path),
            "key_path": str(key_path),
            "expires_at": expiration,
            "is_valid": is_valid,
        }

    async def install_root_ca(self) -> Dict[str, any]:
        """Install Root CA to system trust store"""
        import platform

        system = platform.system()

        if system == "Darwin":  # macOS
            return await self._install_ca_macos()
        elif system == "Windows":
            return await self._install_ca_windows()
        elif system == "Linux":
            return await self._install_ca_linux()
        else:
            return {"success": False, "error": f"Unsupported platform: {system}"}

    async def _install_ca_macos(self) -> Dict[str, any]:
        """Install Root CA on macOS"""
        process = await asyncio.create_subprocess_exec(
            "sudo",
            "security",
            "add-trusted-cert",
            "-d",
            "-r",
            "trustRoot",
            "-k",
            "/Library/Keychains/System.keychain",
            str(self.ca_cert_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            return {"success": False, "error": stderr.decode()}

        return {"success": True}

    async def _install_ca_windows(self) -> Dict[str, any]:
        """Install Root CA on Windows"""
        process = await asyncio.create_subprocess_exec(
            "certutil",
            "-addstore",
            "-f",
            "Root",
            str(self.ca_cert_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            return {"success": False, "error": stderr.decode()}

        return {"success": True}

    async def _install_ca_linux(self) -> Dict[str, any]:
        """Install Root CA on Linux"""
        # This would vary by distribution
        return {"success": False, "error": "Linux installation not implemented yet"}

    async def check_root_ca_status(self) -> Dict[str, any]:
        """Check if Root CA is installed and trusted"""
        import platform

        system = platform.system()

        if not self.ca_cert_path.exists():
            return {"installed": False, "valid": False}

        if system == "Darwin":  # macOS
            return await self._check_ca_status_macos()
        elif system == "Windows":
            return await self._check_ca_status_windows()
        elif system == "Linux":
            return await self._check_ca_status_linux()
        else:
            return {"installed": False, "valid": False}

    async def _check_ca_status_macos(self) -> Dict[str, any]:
        """Check Root CA status on macOS"""
        process = await asyncio.create_subprocess_exec(
            "security",
            "find-certificate",
            "-c",
            "DJANGOForge",
            "/Library/Keychains/System.keychain",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()

        installed = process.returncode == 0

        # Check if trusted
        process = await asyncio.create_subprocess_exec(
            "security",
            "verify-cert",
            "-c",
            str(self.ca_cert_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()

        valid = process.returncode == 0

        return {"installed": installed, "valid": valid}

    async def _check_ca_status_windows(self) -> Dict[str, any]:
        """Check Root CA status on Windows"""
        process = await asyncio.create_subprocess_exec(
            "certutil",
            "-store",
            "Root",
            "DJANGOForge",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()

        installed = process.returncode == 0

        return {"installed": installed, "valid": installed}

    async def _check_ca_status_linux(self) -> Dict[str, any]:
        """Check Root CA status on Linux"""
        return {"installed": False, "valid": False}

File: core/src/privilege/certificate.rs
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CertificateInfo {
    pub domain: String,
    pub certificate_path: String,
    pub key_path: String,
    pub expires_at: String,
    pub is_valid: bool,
}

pub async fn generate_certificate(domain: &str) -> Result<CertificateInfo, String> {
    let djamp_home = dirs::home_dir()
        .ok_or("Home directory not found")?
        .join(".djamp");
    
    let ca_dir = djamp_home.join("ca");
    let cert_dir = djamp_home.join("certs");
    
    let ca_cert = ca_dir.join("djamp-root-ca.crt");
    let ca_key = ca_dir.join("djamp-root-ca.key");
    
    if !ca_cert.exists() || !ca_key.exists() {
        return Err("Root CA not found. Please install Root CA first.".to_string());
    }
    
    let cert_key = cert_dir.join(format!("{}.key", domain));
    let cert_crt = cert_dir.join(format!("{}.crt", domain));
    
    fs::create_dir_all(&cert_dir).map_err(|e| format!("Failed to create cert directory: {}", e))?;
    
    // Generate private key
    Command::new("openssl")
        .args(["genrsa", "-out", cert_key.to_str().unwrap(), "2048"])
        .output()
        .map_err(|e| format!("Failed to generate private key: {}", e))?;
    
    // Generate CSR
    let config_file = cert_dir.join(format!("{}.conf", domain));
    let config_content = format!(
        r#"[req]
distinguished_name = req_distinguished_name
req_extensions = v3_req
prompt = no

[req_distinguished_name]
CN = {}

[v3_req]
keyUsage = keyEncipherment, dataEncipherment
extendedKeyUsage = serverAuth
subjectAltName = @alt_names

[alt_names]
DNS.1 = {}
DNS.2 = www.{}
"#,
        domain, domain, domain
    );
    
    std::fs::write(&config_file, config_content)
        .map_err(|e| format!("Failed to write config: {}", e))?;
    
    Command::new("openssl")
        .args([
            "req",
            "-new",
            "-key",
            cert_key.to_str().unwrap(),
            "-out",
            cert_dir.join(format!("{}.csr", domain)).to_str().unwrap(),
            "-config",
            config_file.to_str().unwrap(),
        ])
        .output()
        .map_err(|e| format!("Failed to generate CSR: {}", e))?;
    
    // Sign with CA
    Command::new("openssl")
        .args([
            "x509",
            "-req",
            "-in",
            cert_dir.join(format!("{}.csr", domain)).to_str().unwrap(),
            "-CA",
            ca_cert.to_str().unwrap(),
            "-CAkey",
            ca_key.to_str().unwrap(),
            "-CAcreateserial",
            "-out",
            cert_crt.to_str().unwrap(),
            "-days",
            "365",
            "-sha256",
            "-extensions",
            "v3_req",
            "-extfile",
            config_file.to_str().unwrap(),
        ])
        .output()
        .map_err(|e| format!("Failed to sign certificate: {}", e))?;
    
    // Clean up
    std::fs::remove_file(cert_dir.join(format!("{}.csr", domain))).ok();
    std::fs::remove_file(&config_file).ok();
    
    // Get expiration date
    let output = Command::new("openssl")
        .args(["x509", "-enddate", "-noout", "-in", cert_crt.to_str().unwrap()])
        .output()
        .map_err(|e| format!("Failed to get certificate expiration: {}", e))?;
    
    let expires_at = String::from_utf8_lossy(&output.stdout)
        .to_string()
        .replace("notAfter=", "")
        .trim()
        .to_string();
    
    Ok(CertificateInfo {
        domain: domain.to_string(),
        certificate_path: cert_crt.to_str().unwrap().to_string(),
        key_path: cert_key.to_str().unwrap().to_string(),
        expires_at,
        is_valid: true,
    })
}

pub async fn check_certificate_status(domain: &str) -> Result<CertificateInfo, String> {
    let djamp_home = dirs::home_dir()
        .ok_or("Home directory not found")?
        .join(".djamp");
    
    let cert_dir = djamp_home.join("certs");
    let cert_crt = cert_dir.join(format!("{}.crt", domain));
    let cert_key = cert_dir.join(format!("{}.key", domain));
    
    if !cert_crt.exists() || !cert_key.exists() {
        return Ok(CertificateInfo {
            domain: domain.to_string(),
            certificate_path: cert_crt.to_str().unwrap().to_string(),
            key_path: cert_key.to_str().unwrap().to_string(),
            expires_at: String::new(),
            is_valid: false,
        });
    }
    
    // Check validity
    let output = Command::new("openssl")
        .args(["x509", "-checkend", "0", "-in", cert_crt.to_str().unwrap()])
        .output()
        .map_err(|e| format!("Failed to check certificate: {}", e))?;
    
    let is_valid = output.status.success();
    
    // Get expiration date
    let output = Command::new("openssl")
        .args(["x509", "-enddate", "-noout", "-in", cert_crt.to_str().unwrap()])
        .output()
        .map_err(|e| format!("Failed to get certificate expiration: {}", e))?;
    
    let expires_at = String::from_utf8_lossy(&output.stdout)
        .to_string()
        .replace("notAfter=", "")
        .trim()
        .to_string();
    
    Ok(CertificateInfo {
        domain: domain.to_string(),
        certificate_path: cert_crt.to_str().unwrap().to_string(),
        key_path: cert_key.to_str().unwrap().to_string(),
        expires_at,
        is_valid,
    })
}

pub async fn install_root_ca() -> Result<(), String> {
    let djamp_home = dirs::home_dir()
        .ok_or("Home directory not found")?
        .join(".djamp");
    
    let ca_dir = djamp_home.join("ca");
    let ca_cert = ca_dir.join("djamp-root-ca.crt");
    let ca_key = ca_dir.join("djamp-root-ca.key");
    
    // Generate CA if it doesn't exist
    if !ca_cert.exists() || !ca_key.exists() {
        generate_root_ca(&ca_cert, &ca_key)?;
    }
    
    #[cfg(target_os = "macos")]
    {
        // Install to macOS Keychain
        let output = Command::new("sudo")
            .args([
                "security",
                "add-trusted-cert",
                "-d",
                "-r",
                "trustRoot",
                "-k",
                "/Library/Keychains/System.keychain",
                ca_cert.to_str().unwrap(),
            ])
            .output()
            .map_err(|e| format!("Failed to install Root CA: {}", e))?;
        
        if !output.status.success() {
            return Err(format!(
                "Failed to install Root CA: {}",
                String::from_utf8_lossy(&output.stderr)
            ));
        }
    }
    
    #[cfg(windows)]
    {
        // Install to Windows Certificate Store
        let output = Command::new("certutil")
            .args([
                "-addstore",
                "-f",
                "Root",
                ca_cert.to_str().unwrap(),
            ])
            .output()
            .map_err(|e| format!("Failed to install Root CA: {}", e))?;
        
        if !output.status.success() {
            return Err(format!(
                "Failed to install Root CA: {}",
                String::from_utf8_lossy(&output.stderr)
            ));
        }
    }
    
    Ok(())
}

pub async fn check_root_ca_status() -> Result<(bool, bool), String> {
    let djamp_home = dirs::home_dir()
        .ok_or("Home directory not found")?
        .join(".djamp");
    
    let ca_dir = djamp_home.join("ca");
    let ca_cert = ca_dir.join("djamp-root-ca.crt");
    
    if !ca_cert.exists() {
        return Ok((false, false));
    }
    
    #[cfg(target_os = "macos")]
    {
        let output = Command::new("security")
            .args([
                "find-certificate",
                "-c",
                "DJANGOForge",
                "/Library/Keychains/System.keychain",
            ])
            .output()
            .map_err(|e| format!("Failed to check Root CA: {}", e))?;
        
        let installed = output.status.success();
        
        // Check if trusted
        let output = Command::new("security")
            .args([
                "verify-cert",
                "-c",
                ca_cert.to_str().unwrap(),
            ])
            .output()
            .map_err(|e| format!("Failed to verify Root CA: {}", e))?;
        
        let valid = output.status.success();
        
        return Ok((installed, valid));
    }
    
    #[cfg(windows)]
    {
        let output = Command::new("certutil")
            .args([
                "-store",
                "Root",
                "DJANGOForge",
            ])
            .output()
            .map_err(|e| format!("Failed to check Root CA: {}", e))?;
        
        let installed = output.status.success();
        
        return Ok((installed, installed));
    }
    
    #[cfg(not(any(target_os = "macos", windows)))]
    {
        Ok((false, false))
    }
}

fn generate_root_ca(ca_cert: &Path, ca_key: &Path) -> Result<(), String> {
    let ca_dir = ca_cert.parent().ok_or("Invalid CA path")?;
    
    fs::create_dir_all(ca_dir).map_err(|e| format!("Failed to create CA directory: {}", e))?;
    
    Command::new("openssl")
        .args([
            "req",
            "-x509",
            "-newkey",
            "rsa:4096",
            "-keyout",
            ca_key.to_str().unwrap(),
            "-out",
            ca_cert.to_str().unwrap(),
            "-days",
            "3650",
            "-nodes",
            "-subj",
            "/C=US/ST=State/L=City/O=DJANGOForge/OU=Development/CN=DJANGOForge Root CA",
            "-addext",
            "basicConstraints=critical,CA:TRUE,pathlen:0",
            "-addext",
            "keyUsage=critical,keyCertSign,cRLSign",
        ])
        .output()
        .map_err(|e| format!("Failed to generate Root CA: {}", e))?;
    
    Ok(())
}
